lib/spack/spack/mixins.py

```diff
@@ -0,0 +1,177 @@
+            phase = getattr(cls, attr_name)
+
+            # Due to MRO, we may have taken a method from a parent class
+            # and modifying it may influence other packages in unwanted
+            # manners. Solve the problem by copying the phase into the most
+            # derived class.
+            setattr(cls, attr_name, phase.copy())
+
+        # Insert the methods in the appropriate position
+        # in the installation pipeline.
+
+        for phase in PackageMixinsMeta._add_method_before:
+
+            attr_name = attr_fmt.format(phase)
+            phase_obj = getattr(cls, attr_name)
+            fn_list = PackageMixinsMeta._add_method_after[phase]
+
+            for f in fn_list:
+                phase_obj.run_before.append(f)
+
+        # Flush the dictionary for the next class
+        PackageMixinsMeta._add_method_before.clear()
+
+        for phase in PackageMixinsMeta._add_method_after:
+
+            attr_name = attr_fmt.format(phase)
+            phase_obj = getattr(cls, attr_name)
+            fn_list = PackageMixinsMeta._add_method_after[phase]
+
+            for f in fn_list:
+                phase_obj.run_after.append(f)
+
+        # Flush the dictionary for the next class
+        PackageMixinsMeta._add_method_after.clear()
+
+        super(PackageMixinsMeta, cls).__init__(name, bases, attr_dict)
+
+
+def filter_compiler_wrappers(*files, **kwargs):
+    """Substitutes any path referring to a Spack compiler wrapper with the
+    path of the underlying compiler that has been used.
+
+    If this isn't done, the files will have CC, CXX, F77, and FC set to
+    Spack's generic cc, c++, f77, and f90. We want them to be bound to
+    whatever compiler they were built with.
+
+    Args:
+        *files: files to be filtered
+        **kwargs: at present supports the keyword 'after' to specify after
+            which phase the files should be filtered (defaults to 'install').
+    """
+    after = kwargs.get('after', 'install')
+
+    def _filter_compiler_wrappers_impl(self):
+
+        tty.debug('Filtering compiler wrappers: {0}'.format(files))
+
+        # Compute the absolute path of the files to be filtered and
+        # remove links from the list.
+        abs_files = llnl.util.filesystem.find(self.prefix, files)
+        abs_files = [x for x in abs_files if not os.path.islink(x)]
+
+        kwargs = {'ignore_absent': True, 'backup': False, 'string': True}
+
+        x = llnl.util.filesystem.FileFilter(*abs_files)
+
+        x.filter(os.environ['CC'], self.compiler.cc, **kwargs)
+        x.filter(os.environ['CXX'], self.compiler.cxx, **kwargs)
+        x.filter(os.environ['F77'], self.compiler.f77, **kwargs)
+        x.filter(os.environ['FC'], self.compiler.fc, **kwargs)
+
+        # Remove this linking flag if present (it turns RPATH into RUNPATH)
+        x.filter('-Wl,--enable-new-dtags', '', **kwargs)
+
+    PackageMixinsMeta.register_method_after(
+        _filter_compiler_wrappers_impl, after
```

Comment:
Could the code in `PackageMixinsMeta` and the mental requirements on mixin implementors be reduced by making this a bit more declarative?

Whoever writes a mixin has to know about `PackageMixinsMeta` as well as the methods on it, which do fairly boring stuff (add before, add after, etc).  What if they just wrote, say:

```python
@mixin
def filter_compiler_wrappers(*files, after='install'):
    def blah():
        # ... etc ...
    return {'after': {'install': {blah}}}
```

Then let the mixin decorator do its magic on the dict there.  Thoughts?  Seems easier to write docs for, in case we write docs for mixin function implementors.